### ISSUE TITLE
Need to upgrade Bouncy Castle dependency bcprov-jdk15on

### ISSUE THREAD

**Issue by malcomm** [DirectProject/nhin-d#263]
- **Agent Version:** 2.1.1
- **Gateway Version:** 4.1

At the moment nhin-d code relies on Bouncy Castle:

```
        <dependency>
            <groupId>bouncycastle</groupId>
            <artifactId>bcprov-jdk15</artifactId>
            <version>140</version>    
        </dependency>   
        <dependency>
            <groupId>bouncycastle</groupId>
            <artifactId>bcmail-jdk15</artifactId>
            <version>140</version>      
        </dependency>   
```

This is from 2008 - [bcprov-jdk15](http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22bouncycastle%22%20AND%20a%3A%22bcprov-jdk15%22) and we are hitting a Bouncy Castle thread issue for **certificate revocation**. Here's the stacktrace:

```
java.lang.IllegalArgumentException: Bad sequence size: 3
at org.bouncycastle.asn1.x509.TBSCertList$CRLEntry.<init>(Unknown Source)
at org.bouncycastle.asn1.x509.TBSCertList.getRevokedCertificates(Unknown Source)
at org.bouncycastle.asn1.x509.CertificateList.getRevokedCertificates(Unknown Source)
at org.bouncycastle.jce.provider.X509CRLObject.isRevoked(Unknown Source)
at BouncyCastleTest.run(BouncyCastleTest.java:55)
```

We were able to isolate this to a separate unit test and swapping out to the latest Bouncy Castle fixed this issue:

```
        <dependency>
            <groupId>org.bouncycastle</groupId>
            <artifactId>bcprov-jdk15on</artifactId>
            <version>1.55</version>
        </dependency>
```

The problem: we are unable to upgrade to this version because nhin-d requires the older version (at least this is my current understanding - please tell me if I'm wrong here [I would be very happy if I am wrong]).

**Comment by gm2552** [comment]
Looking into this.  Updating to later versions of the BC libraries is not trivial due to observed non-passive changes in the public APIs.  Agreed that it is about time to pay some technical debt as the currently used libraries are approaching (if not past) 5 years old.

**Comment by malcomm** [comment]
Agreed, that this is not trivial. I was looking at doing this locally and realized more than a few things have changed. I would be happy to assist with work. Please let me know.

Thanks for looking into this.

**Comment by gm2552** [comment]
I am familiar with this issue and it was resolved in the 2.0.x branch in version 2.0.16.  The code has been moved forward into the trunk branch and will be released with version 2.2 which will go into release next week.

**Comment by malcomm** [comment]
Hi Greg - this sounds great. I see that agent 2.2 is out now. Just to confirm, 2.2 has this fix? Also, do you have an issue/ticket that addresses this?

**Comment by gm2552** [comment]
Yes, version 2.2 has the fix.

The original fix came out in the 2.0.x branch, and the release[1] notes show it was added in 2.0.16.  The 2.1.x branch is not being further supported, so it did not receive this fix.  

Bug 257 [2] is related to this issue, but does not directly fix the issue.  However bug 257 and the threading fix were released together.

[1] http://api.directproject.info/agent/2.2/releaseNotes.html
[2] https://github.com/DirectProject/nhin-d/issues/257

**Comment by gm2552** [comment]
FYI, The modernization project of Java RI 6.0 has upgraded to the latest version of bcprov-jdk15on.
